File: damon/src/main/java/com/zyhang/damon/ParcelFn.java
package com.zyhang.damon;

import android.os.Parcel;

/**
 * ProjectName:Damon
 * Description:
 * Created by zyhang on 2017/4/28.22:34
 * Modify by:
 * Modify time:
 * Modify remark:
 */

final class ParcelFn {
    private static final ClassLoader CLASS_LOADER = ParcelFn.class.getClassLoader();

    static <T> T unmarshall(byte[] array) {
        Parcel parcel = Parcel.obtain();
        parcel.unmarshall(array, 0, array.length);
        parcel.setDataPosition(0);
        Object value = parcel.readValue(CLASS_LOADER);
        parcel.recycle();
        //noinspection unchecked
        return (T) value;
    }

    static byte[] marshall(Object o) {
        Parcel parcel = Parcel.obtain();
        parcel.writeValue(o);
        byte[] result = parcel.marshall();
        parcel.recycle();
        return result;
    }
}

File: damon-kotlin/src/main/kotlin/com/zyhang/damon/kotlin/ParcelFn.kt
package com.zyhang.damon.kotlin

import android.os.Parcel

/**
 * ProjectName:Damon
 * Description:
 * Created by zyhang on 2017/4/28.下午5:43
 * Modify by:
 * Modify time:
 * Modify remark:
 */

internal object ParcelFn {
    val CLASS_LOADER: ClassLoader = ParcelFn::class.java.classLoader

    fun <T> unmarshall(array: ByteArray): T {
        val parcel = Parcel.obtain()
        parcel.unmarshall(array, 0, array.size)
        parcel.setDataPosition(0)
        val value = parcel.readValue(CLASS_LOADER)
        parcel.recycle()
        @Suppress("UNCHECKED_CAST")
        return value as T
    }

    fun marshall(o: Any): ByteArray {
        val parcel = Parcel.obtain()
        parcel.writeValue(o)
        val result = parcel.marshall()
        parcel.recycle()
        return result
    }
}